File: leetcode/89_GrayCode.java
class Solution {
    public List<Integer> grayCode(int n) {
        List<Integer> codes = new ArrayList<>();
        Set<Integer> done = new HashSet<>();
        codes.add(0);
        done.add(0);
        for (int j = 1; j < Math.pow(2, n); j++) {
            int prev = codes.get(codes.size() - 1);
            for (int i = 0; i < n; i++) {
                int cand = prev ^ (1 << i);
                if (done.contains(cand)) continue;
                codes.add(cand);
                done.add(cand);
                break;
            }
        }
        return codes;
    }
}

File: leetcode/go/89_gray_code.go
func grayCode(n int) []int {
	a := make([]int, 1)

	for i := 0; i < n; i++ {
		currSize := len(a)

		for j := currSize - 1; j >= 0; j-- {
			a = append(a, a[j] | 1 << i)
		}
	}

	return a
}
